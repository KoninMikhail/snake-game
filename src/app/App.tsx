import { useEffect } from 'react';
import { useEffectOnce, useUpdateEffect } from 'usehooks-ts';
import { Outlet, ScrollRestoration, useLocation } from 'react-router-dom';
import { FontStyles } from '@styles/fonts-face';
import { GlobalStyles } from '@styles/globals';
import { appActions } from '@store/slices/app.slice';
import { startAppListening } from '@store/store';
import useWindowOrientation from '@hooks/useWindowOrientation/useWindowOrientation';
import useDeviceScreenType from '@hooks/useDeviceScreenType/useDeviceScreenType';
import useActionCreators from '@hooks/useActionCreators/useActionCreators';
import { setupAppListeners, setupGameListeners } from '@store/middlewares/setupListeners';
import { Loader } from '@ui/atoms/loaders/Loader/Loader';
import { isTouchDevice } from '@helpers/validators/isTouchDevice';

/**
 * App
 */
const App = () => {
    const actions = useActionCreators(appActions);
    const location = useLocation();

    /** Detect Screen type
     * ======================== */
    const screen = useDeviceScreenType();
    useUpdateEffect(() => {
        actions.setDeviceScreenType({ screenType: screen.type });
    }, [screen]);

    /** Detect Orientation
     * ======================== */
    const { orientation } = useWindowOrientation();
    useEffect(() => {
        actions.setDeviceOrientation({ orientation });
    }, [actions, orientation]);

    /** Detect touch
     * ======================== */
    useEffectOnce(() => {
        isTouchDevice() && actions.setDeviceTouchState({ touch: true });
    });

    /** Detect Change Route
     * ======================== */
    useEffect(() => {
        actions.setRoute({ route: location.pathname });
    }, [actions, location.pathname]);

    /** Middlewares
     * ======================== */
    useEffectOnce(() => {
        setupAppListeners(startAppListening);
        return () => setupGameListeners(startAppListening)();
    });
    return (
        <>
            <GlobalStyles />
            <FontStyles />
            <Loader />
            <Outlet />
            <ScrollRestoration />
        </>
    );
};

export default App;
