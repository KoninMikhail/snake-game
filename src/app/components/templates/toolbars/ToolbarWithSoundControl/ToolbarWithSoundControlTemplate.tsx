import { SoundControl } from '@ui/organisms/SoundControl/SoundControl';
import { ToolbarWithSoundControlContainer } from '@templates/toolbars/ToolbarWithSoundControl/styles/StyledToolbarWithSoundControl';
import { Column } from '@templates/grids/Row/Column/Column';
import { Row } from '@templates/grids/Row/Row';

export const ToolbarWithSoundControlTemplate = () => (
    <>
        <ToolbarWithSoundControlContainer>
            <Row>
                <Column align={'center'} justify={'end'}>
                    <SoundControl size={'xl'} />
                </Column>
            </Row>
        </ToolbarWithSoundControlContainer>
    </>
);
