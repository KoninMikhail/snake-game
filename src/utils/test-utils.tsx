import React, {PropsWithChildren} from 'react';
import type {RenderOptions} from '@testing-library/react';
import {render} from '@testing-library/react';
import type {PreloadedState} from '@reduxjs/toolkit';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import type {AppStore, RootState} from '@store/store';
import appSlice from '@store/slices/app.slice';
import soundSlice from '@store/slices/sound.slice';
import maxScoreSlice from '@store/slices/maxScore.slice';
import gameDataSlice from '@store/slices/game.slice';
import {ImagesData} from '@data/imagesData';
import {mockImagesData, mockSoundData} from '@/setupTests';
import SoundsProvider from '@context/SoundsContext';
import ImagesProvider from '@context/ImagesContext';
import {SoundsData} from '@data/soundsData';
import {DefaultTheme, ThemeProvider} from 'styled-components';
import {theme as appTheme} from '@styles/theme';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
    images?: ImagesData;
    sounds?: SoundsData;
    theme?: DefaultTheme;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                app: appSlice,
                sounds: soundSlice,
                maxScore: maxScoreSlice,
                game: gameDataSlice,
            },
            preloadedState,
        }),
        images = mockImagesData,
        sounds = mockSoundData,
        theme = appTheme,
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({children}: PropsWithChildren): JSX.Element {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <SoundsProvider sounds={sounds}>
                        <ImagesProvider images={images}>{children}</ImagesProvider>
                    </SoundsProvider>
                </ThemeProvider>
            </Provider>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
