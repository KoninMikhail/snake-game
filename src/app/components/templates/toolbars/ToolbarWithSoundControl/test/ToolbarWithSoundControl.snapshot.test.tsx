import { ToolbarWithSoundControlTemplate } from '@templates/toolbars/ToolbarWithSoundControl/ToolbarWithSoundControlTemplate';
import { renderWithProviders } from '@utils/test-utils';
import { AppGlobalState } from '@store/slices/app.slice';

/**
 * Mocks
 */

// fake content
const mockSoundContent = 'mocked SoundControl';

// fake component
jest.mock('@ui/organisms/SoundControl/SoundControl', () => ({
    SoundControl: () => mockSoundContent,
}));

/**
 * Tests
 */
describe('ToolbarWithSoundControl', () => {
    it('Equals snapshot on small screens', () => {
        const preloadedState = {
            preloadedState: { app: { device: { screen: 'small' } } as AppGlobalState },
        };

        const { container } = renderWithProviders(
            <ToolbarWithSoundControlTemplate />,
            preloadedState
        );

        expect(container).toMatchSnapshot();
    });

    it('Equals snapshot on middle screen devices', () => {
        const preloadedState = {
            preloadedState: { app: { device: { screen: 'middle' } } as AppGlobalState },
        };

        const { container } = renderWithProviders(
            <ToolbarWithSoundControlTemplate />,
            preloadedState
        );

        expect(container).toMatchSnapshot();
    });
    it('Equals snapshot on large screen devices', () => {
        const preloadedState = {
            preloadedState: { app: { device: { screen: 'large' } } as AppGlobalState },
        };

        const { container } = renderWithProviders(
            <ToolbarWithSoundControlTemplate />,
            preloadedState
        );

        expect(container).toMatchSnapshot();
    });
});
