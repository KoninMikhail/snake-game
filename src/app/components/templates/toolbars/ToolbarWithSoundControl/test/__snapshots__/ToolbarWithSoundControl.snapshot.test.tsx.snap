// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarWithSoundControl Equals snapshot on large screen devices 1`] = `
.c0 {
  display: block;
  width: 100%;
}

.c2 {
  display: block;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: end;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

<div>
  <header
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        mocked SoundControl
      </div>
    </div>
  </header>
</div>
`;

exports[`ToolbarWithSoundControl Equals snapshot on middle screen devices 1`] = `
.c0 {
  display: block;
  width: 100%;
}

.c2 {
  display: block;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: end;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

<div>
  <header
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        mocked SoundControl
      </div>
    </div>
  </header>
</div>
`;

exports[`ToolbarWithSoundControl Equals snapshot on small screens 1`] = `
.c0 {
  display: block;
  width: 100%;
}

.c2 {
  display: block;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: end;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

<div>
  <header
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        mocked SoundControl
      </div>
    </div>
  </header>
</div>
`;
